### DISCUSSION TITLE
Is the libaray support the parralel

### DISCUSSION THREAD

**Discussion by Hubuwenxin** [Rubberazer/JETGPIO#26] [Q&A] [answered]
![Screenshot from 2024-05-08 10-10-08](https://github.com/Rubberazer/JETGPIO/assets/132553604/f2f0708d-fcba-40db-83ab-e4ec35d9927d)
as shown in the picture, there is two type where are AD7606  SPI and parallel in the jetson agx orin. i notice that in the libarary only have SPI examle code, and the function such as spiopen and i2c etc., but i didn't find the parallel function or example, so i want ask that can the libary support the parallel communication here?

**Comment by Rubberazer** [comment, marked as answer]
Clear, so the question is how to get data from an AD7606 analogue converter into the Orin AGX via parallel port (kind of), the answer is: it is doable but I don't recommend it,  with or without my library, I would recommend to use the SPI interface. So the situation would be something like this below, where the "digital host" is the Orin AGX:
![image](https://github.com/Rubberazer/JETGPIO/assets/47650457/6bb5936a-c6d4-4830-a141-0d784728643b)

The Orin machines (all models) do not have some kind of "parallel" interface per se, you would need to bit bang/create that interface yourself, you would need 16 pins for data + 2 for control (busy & CS/RD). In order to bit bang that, you have to sample each pin individually and compose a word with the sample from the AD7606, it would be something like:

Read interrupt pin (busy pin)--> if Interrupt is low/0 --> read pins AD 16 to 33

This can be done with my library (if my library supported the AGX)? yes, but the pins would be read sequentially, (data pin16 to pin33), would they be read fast enough? Maybe, If the AD7606 keeps the latest sample value for a few us, then yes, reading the AD7606 manual it seems it does but I never played with this AD so not sure, I can't confirm it would be OK.

Summary:

1. My library doesn't support the orin AGX at the moment, only the Nano classic, Orin Nano and Orin NX so it is not going to work in the AGX anyways.
2. In theory it could be done with my library (or others) if my library supported AGX but I recommend the SPI port option
3. Even through SPI, my library is not gonna give you an easy API to get data from the AD7606, it only gives you raw SPI communication, everything else would have to be built on top by you. I like this kind of stuff but many people find it annoying and is time consuming.
4.  Since you show a screenshot of the kernel menuconfig, maybe you could try to do it "the kernel way" and try to find information in how to get the data through the SPI option, probably the kernel will create some kind of character device and you can get data through ioctls or something, not sure if you have tried that already.

Finally, probably you are already aware but there are alternatives to my library, see this one for python: https://github.com/NVIDIA/jetson-gpio or this one for c++: https://github.com/pjueon/JetsonGPIO both support the AGX Orin but none of them is going to give you an easy/ready to go  way of getting data from the AD7606. The main difference between those and mine is that mine is faster :)

**Reply by Hubuwenxin** [reply]
Thank you for reply. Auctually i tried to connect the AD7606 and AGX orin and seccessed to get the data based on your libarary by SPI. The problem is i need to get the data much more faster right now to match some standard, so i try to think about the parrallel communication. I will try to test and tell the result later. 
 i will appreciate it  if you have the experience for improve the data transmit speed.

**Reply by Rubberazer** [reply]
Yes it can be done, but remember that the header pinout of the AGX and the other Orins is different and that is why my library doesn't support the AGX Orin, you would need to modify my library, I would do it but i don't have the AGX to test the changes.
What kind of sampling rate are you trying to achieve?

**Reply by Hubuwenxin** [reply]
The pin for SPI is same with the orin nx, and i defined some GPIO into busy etc., right now it's work well and the data is correct.
Additionally, right now the frequency is 20kHZ but i want to up to 40kHZ.

**Reply by Rubberazer** [reply]
Cool, could you please make me a massive favour and tell me what is working and what is not on the AGX? e.g. pin 3 works fine as output or whatever but 7 doesn't, when you have the time, again I don't have the AGX so I am not able to do it myself now.
Later on I will be creating some program that I think is going to be useful for what you want to do, I have been also thinking in adding a new function that would be helpful.

**Reply by Rubberazer** [reply]
I created this little program [read_ADC7606.c](https://github.com/Rubberazer/JETGPIO/blob/main/EXAMPLES_C/read_ADC7606_v1.c). It could be improved and there are other ways to do it too, but is doing the basic reading cycle, again I don't have the ADC7606 so maybe you will have to do some little changes or so but I believe the essential stuff is there. 
At 40kHz you need a sample every 25us, the reading cycle of this program will execute approximately in 20us.

**Reply by Hubuwenxin** [reply]
Thanks. I tried to run it and changed a few pin. The speed is twice time faster than before but the voltage value is not right. Maybe need to check the GPIO pins or something else.

**Reply by Rubberazer** [reply]
As commented before, I believe the header pinout in the Orin AGX is different from the header pinout in the Orin Nano or NX, so probably my library is not going to work 100% in the AGX, I believe some GPIO pins will work but others will not. I can't confirm any of this because I don't have the AGX.
Sorry but I don't think I can't do much more at the moment until I get my hands onto the AGX model.

**Reply by Hubuwenxin** [reply]
hello, i check the GPIO pins and I have one more question. There is "Reads the GPIO level, on or off, 0 or 1" in C libirary. When i print the pin_level, there is
pin_level 0: 0
pin_level 1: 1
pin_level 2: 1
pin_level 3: 0
pin_level 4: 1
pin_level 5: 1
pin_level 6: 1
pin_level 7: 1
pin_level 8: 0
pin_level 9: 1
pin_level 10: 1
pin_level 11: 0
pin_level 12: 0
pin_level 13: 0
pin_level 14: 0
pin_level 15: 0 
As i know that every output pin of ad7606's is 8bits(4-bit binary) , but It's sounds like here every pin just output 1-bit binary.  Is there any method can change that not receive the 0 and 1 in every pins but every pin is 16 bits?   
i'm looking forward to your reply!

**Reply by Rubberazer** [reply]
No, that's not how the ADC7606 works,

In parallel mode each ADC7606 output from DB0 to DB15 will be 1 bit, logic 0 or 1. That' s why is parallel, when you put all of them together you will get a 16 bit word which is your measured value, for instance with what you are showing me above and assuming pin_level15 is the MSB (DB15 in the ADC7606), the value you are getting is: 0000011011110110 = 1782, then you apply the conversion equation to that and you should get your V value.

**Reply by Hubuwenxin** [reply]
But here still have some problem. Could you please help me  analyze it?
First question is that when i test on the parallel communication with ad7606, i find that the value of pin_level is different from the what i set in ad7606's input;
Then i test AGX orin's every pin,find that some of the pin writting pin_level 0; but actually the voltage is 3.3V.  And some pin writting pin_level 10: 1; but the voltage is 0V. I feel so confused about this. if you have knowlegement about this, sould you please give me anaysis.

The final question is that there are 4 input channels (every channel is 16bit) in ad7606, it dose matter i can get the whole 4 channel?

**Reply by Rubberazer** [reply]
I told you many times, my library does NOT support the Orin AGX model, only the Orin Nano and Orin NX, so some pins will work and some will not, I can't upgrade the library because I don't have the AGX to do it. You will have to do it yourself or find another solution.
Related to your question on the ADC, yes you can get up to 8 values, V1 to V8 or only 4 V1 to V4 for instance, depending on the value of: CONVST A & CONVST B. Each value will show up on the ADC pins every time you toggle pin: CS/RD

**Reply by Hubuwenxin** [reply]
thank your for your kind reply! If i finding the problem i will share it here. 
hope you can success to extend the library   AGX orin .

**Reply by Rubberazer** [reply]
Really sorry I can´t help you more, if you know anybody in a position to send me a spare AGX over here, well, just pass the message if you want! Anybody interested can contact me here: rubberazer@outlook.com
